mm2src/coins/tendermint/tendermint_tx_history_v2.rs

```diff
@@ -0,0 +1,803 @@
+            },
+        };
+
+        let q = format!(
+            "coin_received.receiver = '{}' AND tx.height > {}",
+            self.address.clone(),
+            self.from_block_height
+        );
+        let highest_recieved_tx_height = match fetch_and_insert_txs(
+            &self.tendermint_assets_conf,
+            self.address.clone(),
+            &ctx.coin,
+            &ctx.storage,
+            q,
+            self.from_block_height,
+        )
+        .await
+        {
+            Ok(block) => block,
+            Err(stopped) => {
+                if let StopReason::RpcClient(e) = &stopped.stop_reason {
+                    log::error!("Received tx history process turned into cooldown mode due to rpc error: {e}");
+                    return Self::change_state(OnIoErrorCooldown::new(
+                        self.address.clone(),
+                        self.from_block_height,
+                        self.tendermint_assets_conf,
+                    ));
+                }
+
+                return Self::change_state(stopped);
+            },
+        };
+
+        let last_fetched_block = cmp::max(highest_send_tx_height, highest_recieved_tx_height);
+
+        log::info!(
+            "Tx history fetching finished for tendermint. Last fetched block {}",
+            last_fetched_block
+        );
+
+        ctx.coin.set_history_sync_state(HistorySyncState::Finished);
+        Self::change_state(WaitForHistoryUpdateTrigger::new(
+            self.address.clone(),
+            last_fetched_block,
+            self.tendermint_assets_conf,
+        ))
+    }
+}
+
+#[async_trait]
+impl<Coin, Storage> State for TendermintInit<Coin, Storage>
+where
+    Coin: TendermintTxHistoryOps,
+    Storage: TxHistoryStorage,
+{
+    type Ctx = TendermintTxHistoryCtx<Coin, Storage>;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> StateResult<Self::Ctx, Self::Result> {
+        const INITIAL_SEARCH_HEIGHT: u64 = 0;
+        const COIN_PROTOCOL_TYPE: &str = "TENDERMINT";
+        const TOKEN_PROTOCOL_TYPE: &str = "TENDERMINTTOKEN";
+
+        ctx.coin.set_history_sync_state(HistorySyncState::NotStarted);
+
+        if let Err(e) = ctx.storage.init(&ctx.coin.history_wallet_id()).await {
+            return Self::change_state(Stopped::storage_error(e));
+        }
+
+        let tendermint_assets: Vec<Json> = ctx.mm_ctx.conf["coins"]
+            .as_array()
+            .expect("couldn't read coins context")
+            .iter()
+            .map(|coin| coin.to_owned())
+            .filter(|coin| {
+                matches!(
+                    coin["protocol"]["type"].as_str(),
+                    Some(COIN_PROTOCOL_TYPE) | Some(TOKEN_PROTOCOL_TYPE)
+                )
+            })
+            .collect();
+
+        Self::change_state(FetchingTransactionsData::new(
+            ctx.coin.my_address().expect("my_address can't fail"),
+            INITIAL_SEARCH_HEIGHT,
```

Comment:
I suppose we can fetch the max block height from the DB and use it here. With one note: we can't be sure that we really added all transactions from specific block on previous run, so it's better to start from `max_block_height_from_db - 1`.

Reply:
Done for `non-wasm` targets. For `wasm`, search will start from `0` as it was. It requires quite some work on indexed db side, I think it might be overkill for this PR. I added `TODO` note into the related wasm function.